### ISSUE TITLE
[RFC]:Unified RL Configuration via additional_config.rl_config

### ISSUE THREAD

**Issue by Ronald1995** [vllm-project/vllm-ascend#13706]
## Motivation

Currently, vLLM Ascend users running RL (Reinforcement Learning) workloads must manually configure **6+ scattered options** across 3 different layers:

1. **vLLM upstream configs** (e.g., `enable_sleep_mode`, `enable_return_routed_experts`, `SamplingParams`)
2. **vllm-ascend `additional_config`** (e.g., `weight_nz_mode`, `enable_sleep_mode_extra_cleanup`)
3. **Environment variables** (e.g., `VLLM_ASCEND_ENABLE_NZ`, `VLLM_SERVER_DEV_MODE`, `VLLM_BATCH_INVARIANT`)

Misconfiguring any single item leads to runtime errors (e.g., NZ mode detected at `wake_up()` time rather than startup), wasted debugging cycles, or suboptimal memory usage. This RFC proposes a unified `additional_config.rl_config` dict to provide a **one-click RL mode** with sensible defaults, while preserving override flexibility for advanced users.

---

#### 1.1 RL Deployment Modes

vLLM Ascend supports two RL deployment modes. **Both modes require weight updates via `weight_transfer_config`**, and differ in memory management and the weight transfer backend:

| Deployment Mode | Description | Memory Management | Weight Transfer Backend |
|---|---|---|---|
| **Same-Device Mode** (same-device) | The inference engine and trainer share the same NPU card | `sleep(level)` / `wake_up(tags)` — release/restore NPU memory to the trainer | **IPC** (`"ipc"`) — shared memory between processes on the same card |
| **Cross-Device Mode** (cross-device) | The inference engine and trainer use different NPU cards | `pause_generation(mode)` / `resume_generation()` — only pause/resume scheduling, does not release memory | **HCCL** (`"nccl"`) — cross-card HCCL communication |

#### 1.2 Same-Device Mode (Sleep/Wake + IPC Weight Transfer)

| Config / API | Type | Purpose | Use Case |
|---|---|---|---|
| `enable_sleep_mode` | bool (model_config) | Enables CaMemAllocator memory pool management | Releases NPU memory to the trainer before training |
| `sleep(level=1)` | Engine API | Offloads weights to CPU, discards KV cache | Keeps a weight copy, restores after training completes |
| `sleep(level=2)` | Engine API | Discards weights + KV cache | For model switching or reloading weights |
| `wake_up(tags)` | Engine API | Restores weights / KV cache in stages | Restores inference after training completes |
| `weight_transfer_config` | dict (vllm config) | `{"backend": "ipc"}` — same-device IPC weight synchronization | Pushes updated weights from the trainer to the inference engine |

#### 1.3 Cross-Device Mode (Pause/Resume + HCCL Weight Transfer)

| Config / API | Type | Purpose | Use Case |
|---|---|---|---|
| `weight_transfer_config` | dict (vllm config) | `{"backend": "nccl"}` — cross-card HCCL weight synchronization | Pushes updated weights from the trainer to the inference engine |
| `pause_generation(mode)` | Engine API | Pauses request scheduling without releasing memory (`mode="keep"` freezes requests; `mode="wait"` drains requests then blocks) | Pauses inference during weight synchronization |
| `resume_generation()` | Engine API | Resumes request scheduling | Restores inference after weight synchronization completes |

#### 1.4 Common Weight Transfer API (shared by both modes)

| Config / API | Type | Purpose | Use Case |
|---|---|---|---|
| `init_weight_transfer_engine(info)` | Engine API | Initializes the communication group between the trainer and vLLM (IPC or HCCL) | Before the training loop starts |
| `start_weight_update()` | Engine API | Marks the start of a weight update; internally validates `weight_nz_mode==0` | Before each weight synchronization round |
| `update_weights(info)` | Engine API | Receives and loads weight chunks sent by the trainer | During weight transfer |
| `finish_weight_update()` | Engine API | Marks the weight update complete; performs post-processing (layout restoration, etc.) | After weight synchronization completes |

#### 1.5 RL Inference Features (upstream vLLM, required for RL training)

| Config / API | Type | Purpose | Notes |
|---|---|---|---|
| `--enable-return-routed-experts` | CLI / model_config | **Router Replay (R3)**: records the expert routing decisions made by MoE models during inference | Outputs `CompletionOutput.routed_experts` (shape: `[seq_len, num_moe_layers, top_k]`), which the trainer replays to avoid gradient corruption caused by inference/training expert routing mismatch |
| `SamplingParams(return_prompt_hidden_states=True)` | SamplingParams | **Extract Hidden States**: returns the hidden states of the prompt | Outputs the `prompt_hidden_states` tensor (shape: `[num_prompt_tokens, hidden_size]`), used for value function training in RL or teacher-student knowledge distillation (OPD) |
| `SamplingParams(logprobs=...)` / `prompt_logprobs` | SamplingParams | **Token Logprobs**: returns per-token log probabilities | RL (PPO/GRPO) requires per-token logprobs to compute the advantage and importance sampling ratio |
| `vllm serve-tokens` | CLI (RFC, [#22817](https://github.com/vllm-project/vllm/issues/22817)) | **Token In/Out**: a tokens-only input/output API server that bypasses the OpenAI text layer | RL training frameworks consume the raw token ID stream directly, eliminating text serialization/deserialization overhead. Provides `/generate` and `/generate_streaming` endpoints, with `GenerateRequest` / `GenerateResponse` dataclasses passing `token_ids`, `logprobs`, and `features` tensors directly |
| `VLLM_BATCH_INVARIANT` | env var (bool) | Enables batch-invariant deterministic computation | Ensures batch composition changes do not affect output; used for RL validation |

### 2. vllm-ascend `additional_config` (Ascend-specific RL extensions)

| Config Key | Type | Default | RL Requirement | Description |
|---|---|---|---|---|
| `weight_nz_mode` | int | `1` | **Both modes: Must be `0`** | 0=disable NZ, 1=quant only, 2=all. The FRACTAL_NZ weight layout can cause parameter precision issues after multiple sleep/wake_up or weight_transfer cycles in RL scenarios. Validated at `wake_up()` in same-device mode, and at `start_weight_update()` in cross-device mode. |
| `enable_sleep_mode_extra_cleanup` | bool | `False` | **Same-device mode: Should be `True`** | In same-device mode, additionally releases the HCCL process group and ACL graph attention workspace during sleep to free up more NPU memory for the trainer. Trade-off: increased wakeup latency (HCCL must be rebuilt and graphs recaptured). Not needed in cross-device mode. |
| `multistream_dsv4_dsa_overlap` | bool | `True` | RL note | DeepSeek V4 DSA multi-stream overlap. In RL scenarios the graph is captured only once, so weight addresses must remain stable across wakeup cycles. |
| `refresh` | bool | `False` | Test/debug helper | Forces a rebuild of the `AscendConfig` singleton when the same `VllmConfig` object is reused within one process but `additional_config` has changed. Not needed when creating a new `LLM()` instance (the `is` comparison automatically fails). |

### 3. Environment Variables (managed outside `additional_config`)

| Env Var | Default | RL Requirement | Description |
|---|---|---|---|
| `VLLM_ASCEND_ENABLE_NZ` | `1` | **Both modes: Must be `0`** | Legacy env var for NZ mode (deprecated in favor of `weight_nz_mode`) |
| `VLLM_SERVER_DEV_MODE` | unset | **Online: Must be `1`** | Exposes HTTP endpoints such as `/sleep`, `/wake_up`, `/pause`, and `/resume`. Not needed for the offline `LLM()` API. |
| `PYTORCH_NPU_ALLOC_CONF` | `expandable_segments:True` | **Same-device mode: must not contain `expandable_segments`** | When `enable_sleep_mode=True`, `platform.py` automatically skips setting this (CaMemAllocator is mutually exclusive with expandable_segments). Cross-device mode has no such restriction. |
| `HCCL_DETERMINISTIC` | unset | `strict` (when batch_invariant) | Set automatically by `batch_invariant.py` when `VLLM_BATCH_INVARIANT=1` |
| `LCCL_DETERMINISTIC` | unset | `1` (when batch_invariant) | Set automatically by `batch_invariant.py` when `VLLM_BATCH_INVARIANT=1` |

### 4. Implicit RL Adaptations (code-level, no user config needed)

| Location | Mechanism |
|---|---|
| `platform.py:693-708` | Auto-skips `PYTORCH_NPU_ALLOC_CONF=expandable_segments:True` when `enable_sleep_mode=True` |
| `worker.py:248-253` | `wake_up()` validates `weight_nz_mode==0`, raises `ValueError` otherwise |
| `worker.py:318-324` | `_check_nz_disabled()` called before `start_weight_update()` |
| `worker.py:260-289` | `wake_up()` restores MoE expert weight transpose layout for unquantized models |
| `w8a8_mxfp8.py:107-150` | `restore_weights_for_rl_loading()` / `process_weights_after_loading()` for quantized model weight reload |
| `mla_v1.py:929`, `sfa_v1.py:705` | RL + graph scenario: capture graph once, weight address expected stable |

---

### Design: `additional_config.rl_config` Dict

Introduce an `RlConfig` class that serves as a **one-click RL mode switch**. When `rl_config.enabled=True`, all RL best-practice defaults are applied automatically, and conflicts are detected at config initialization time (fail-fast) rather than at runtime.

The design recognizes the two deployment modes — users select the mode, and `rl_config` applies the appropriate defaults.

### Configuration Schema

```json
{
    "rl_config": {
        "enabled": false,
        "sleep_mode_extra_cleanup": false,
        "weight_nz_mode": 0,
        "disable_expandable_segments": true,
        "enable_training_consistency": false,
        "enable_batch_invariant": false,
        "enable_dev_endpoints": true
    }
}
```

### Field Details

| Field | Type | Default | Description |
|---|---|---|---|
| `enabled` | bool | `false` | Master switch for RL mode. When `true`, applies all RL best-practice defaults. |
| `sleep_mode_extra_cleanup` | bool | `false` | [Same-device mode] Enables HCCL group release + ACL graph workspace cleanup during sleep, returning more NPU memory to the trainer. **Default `false`** — users who need extra memory saving in same-device mode should explicitly opt in, as it increases wakeup latency. Overrides the top-level `additional_config.enable_sleep_mode_extra_cleanup`. |
| `weight_nz_mode` | int | `0` | [Both modes] Disables FRACTAL_NZ weight layout conversion. Applies to both same-device (sleep/wake) and cross-device (weight_transfer) scenarios. Overrides the top-level `additional_config.weight_nz_mode`. Validated at `AscendConfig.__init__` time — conflicts are rejected immediately. |
| `disable_expandable_segments` | bool | `true` | [Same-device mode] Ensures `PYTORCH_NPU_ALLOC_CONF` does not contain `expandable_segments:True`, which is incompatible with CaMemAllocator (sleep mode). No effect in cross-device mode. |
| `enable_training_consistency` | bool | `false` | [Both modes] Enables **training-inference consistency (operator alignment)**. Aligns the inference model's operators with those of the training model, so the inference engine executes the same operators as the training framework and results match training more closely. This is independent of batch consistency. |
| `enable_batch_invariant` | bool | `false` | [Both modes] Enables **batch-invariant deterministic computation** so that changes in batch composition do not affect outputs. When `true`, automatically sets: `VLLM_BATCH_INVARIANT=1`, `HCCL_DETERMINISTIC=strict`, `LCCL_DETERMINISTIC=1`, `torch.use_deterministic_algorithms(True)`, and registers batch-invariant operators (Triton/AscendC). Also enforces `weight_nz_mode=0` since NZ layout breaks determinism. **Requires building vllm-ascend from source with `COMPILE_CUSTOM_KERNELS=1`.** When set through `rl_config`, this takes precedence over environment variables (see Priority Rules). |
| `enable_dev_endpoints` | bool | `true` | [Online mode] Automatically sets `VLLM_SERVER_DEV_MODE=1` for `vllm serve` online mode, exposing `/sleep`, `/wake_up`, `/pause`, `/resume` HTTP endpoints. Respects explicit user override: if `VLLM_SERVER_DEV_MODE=0` is already set, a warning is logged but the user's choice is honored. No effect in offline `LLM()` mode. |

### Priority Rules

```
rl_config sub-fields > additional_config top-level keys > environment variables > code defaults
```

When `rl_config.enabled=True` and a conflict is detected (e.g., `rl_config.weight_nz_mode=0` but top-level `weight_nz_mode=2`), an error is raised at `AscendConfig.__init__` time with a clear message.

Batch consistency is normally controlled through environment variables (`VLLM_BATCH_INVARIANT`, `HCCL_DETERMINISTIC`, `LCCL_DETERMINISTIC`). When `rl_config.enable_batch_invariant` is set, the value in `additional_config` takes precedence over any already-set environment variable, giving a single source of truth.

### User Experience Comparison

**Same-Device Mode (Same-Device RL, IPC Weight Transfer) — Before (current state):**

```bash
VLLM_SERVER_DEV_MODE=1 \
VLLM_ASCEND_ENABLE_NZ=0 \
VLLM_WORKER_MULTIPROC_METHOD=spawn \
VLLM_BATCH_INVARIANT=1 \
vllm serve DeepSeek-V4 \
    --enable-sleep-mode \
    --enable-return-routed-experts \
    --weight-transfer-config '{"backend": "ipc"}' \
    --additional-config '{
        "enable_sleep_mode_extra_cleanup": true,
        "weight_nz_mode": 0
    }'
```

**Same-Device Mode — After (one-click RL mode):**

```bash
VLLM_WORKER_MULTIPROC_METHOD=spawn \
vllm serve DeepSeek-V4 \
    --enable-sleep-mode \
    --enable-return-routed-experts \
    --weight-transfer-config '{"backend": "ipc"}' \
    --additional-config '{"rl_config": {"enabled": true, "enable_training_consistency": true, "enable_batch_invariant": true, "sleep_mode_extra_cleanup": true}}'
```

All of the following are automatically applied:
- ✅ `VLLM_SERVER_DEV_MODE=1`
- ✅ `weight_nz_mode=0`
- ✅ `expandable_segments` disabled
- ✅ Training-inference consistency: inference operators aligned with training operators
- ✅ Batch consistency: `VLLM_BATCH_INVARIANT=1` + `HCCL_DETERMINISTIC=strict` + `LCCL_DETERMINISTIC=1` + deterministic ops + batch-invariant operators

**Cross-Device Mode (Cross-Device RL, HCCL Weight Transfer) — Before (current state):**

```bash
VLLM_SERVER_DEV_MODE=1 \
VLLM_ASCEND_ENABLE_NZ=0 \
VLLM_WORKER_MULTIPROC_METHOD=spawn \
vllm serve DeepSeek-V4 \
    --weight-transfer-config '{"backend": "nccl"}' \
    --additional-config '{"weight_nz_mode": 0}'
```

**Cross-Device Mode — After (one-click RL mode):**

```bash
VLLM_WORKER_MULTIPROC_METHOD=spawn \
vllm serve DeepSeek-V4 \
    --weight-transfer-config '{"backend": "nccl"}' \
    --additional-config '{"rl_config": {"enabled": true}}'
```

All of the following are automatically applied:
- ✅ `VLLM_SERVER_DEV_MODE=1`
- ✅ `weight_nz_mode=0`

**Offline Mode Usage Example:**

```python
llm = LLM(
    model="DeepSeek-V4",
    enable_sleep_mode=True,  # Only needed in same-device mode
    additional_config={
        "rl_config": {
            "enabled": True,
            "enable_training_consistency": True,      # Training-inference consistency
            "enable_batch_invariant": True,          # Batch-invariant determinism
        },
    },
)

sampling_params = SamplingParams(
    temperature=0,
    max_tokens=256,
    logprobs=1,                                 # Token logprobs for RL
    return_prompt_hidden_states=True,            # Extract hidden states
)
```

### Implementation Outline

| File | Change |
|---|---|
| `vllm_ascend/ascend_config.py` | Add `RlConfig` class with defaults, validation, and `_apply()` method. Integrate into `AscendConfig.__init__` after existing config parsing. |
| `vllm_ascend/platform.py` | Extract `expandable_segments` cleanup logic into reusable helper; call from `RlConfig._apply()` |
| `vllm_ascend/worker/worker.py` | Keep runtime NZ checks in `wake_up()` and `_check_nz_disabled()` as defense-in-depth, but add pointer to `rl_config` in error messages |
| `docs/source/user_guide/configuration/additional_config.md` | Add `rl_config` section documenting all fields, two deployment modes, usage examples, and migration guide |
| `tests/ut/test_ascend_config.py` | Add unit tests: default values, conflict detection, override priority, env var interaction |

### Backward Compatibility

- All existing top-level `additional_config` keys (`weight_nz_mode`, `enable_sleep_mode_extra_cleanup`) continue to work unchanged
- Existing environment variables (`VLLM_ASCEND_ENABLE_NZ`, `VLLM_SERVER_DEV_MODE`, etc.) continue to work
- `rl_config` is purely additive — users who don't set it see zero behavior change
- When both `rl_config` and top-level keys are set for the same field, a clear error is raised with guidance to use `rl_config`

---

## Feedback Period

One week (2026-08-06 ~ 2026-08-13).

## CC List

@vllm-project/vllm-ascend-reviewers

## Any Other Things

- `enable_batch_invariant` requires building from source with `COMPILE_CUSTOM_KERNELS=1`. Should `rl_config` detect at init time whether batch-invariant ops are available and give a clear error if not?
- Should `rl_config` also handle `VLLM_WORKER_MULTIPROC_METHOD=spawn` (another commonly-needed RL env var)?
- The `multistream_dsv4_dsa_overlap` RL interaction (graph capture once, weight address stable) is currently implicit. Should it become an explicit `rl_config` field?
- Router Replay (`enable_return_routed_experts`) and Extract Hidden States (`return_prompt_hidden_states`) are upstream vLLM features controlled via CLI / SamplingParams rather than `additional_config`. Should `rl_config` document them as recommended RL settings without directly managing them?
